FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RecipientsViewModel.kt

```diff
@@ -0,0 +1,391 @@
+        .updateSuspend(
+          pgpKeyDetails.toPublicKeyEntity(publicKeyEntity.recipient).copy(
+            id = publicKeyEntity.id,
+            recipient = publicKeyEntity.recipient
+          )
+        )
+    }
+  }
+
+  fun copyPubKeysBetweenRecipients(
+    sourceRecipientEntity: RecipientEntity?,
+    destinationRecipientEntity: RecipientEntity?
+  ) {
+    viewModelScope.launch {
+      sourceRecipientEntity ?: return@launch
+      destinationRecipientEntity ?: return@launch
+
+      val existingPubKeysOfSource =
+        roomDatabase.pubKeyDao().getPublicKeysByRecipient(sourceRecipientEntity.email)
+
+      if (existingPubKeysOfSource.isNotEmpty()) {
+        for (existingPubKey in existingPubKeysOfSource) {
+          val fetchedPubKey = roomDatabase.pubKeyDao().getPublicKeyByRecipientAndFingerprint(
+            destinationRecipientEntity.email,
+            existingPubKey.fingerprint
+          )
+
+          if (fetchedPubKey == null) {
+            roomDatabase.pubKeyDao()
+              .insertSuspend(
+                existingPubKey.copy(
+                  id = null,
+                  recipient = destinationRecipientEntity.email.lowercase()
+                )
+              )
+          }
+        }
+      }
+    }
+  }
+
+  fun filterContacts(searchPattern: String?) {
+    searchPatternLiveData.value = searchPattern
+  }
+
+  fun deleteContactByEmail(email: String) {
+    viewModelScope.launch {
+      roomDatabase.recipientDao().getRecipientByEmailSuspend(email)?.let {
+        roomDatabase.recipientDao().deleteSuspend(it)
+      }
+    }
+  }
+
+  fun fetchPubKeys(keyIdOrEmail: String) {
+    viewModelScope.launch {
+      lookUpPubKeysMutableStateFlow.value = Result.loading()
+      val activeAccount = getActiveAccountSuspend()
+      lookUpPubKeysMutableStateFlow.value =
+        controlledRunnerForPubKeysFromServer.cancelPreviousThenRun {
+          return@cancelPreviousThenRun apiRepository.pubLookup(
+            context = getApplication(),
+            identData = keyIdOrEmail,
+            orgRules = activeAccount?.clientConfiguration
+          )
+        }
+    }
+  }
+
+  private fun setResultForRemoteContactsLiveData(
+    type: RecipientEntity.Type,
+    result: Result<List<RecipientWithPubKeys>>
+  ) {
+    when (type) {
+      RecipientEntity.Type.TO -> {
+        recipientsToLiveData.value = result
+      }
+
+      RecipientEntity.Type.CC -> {
+        recipientsCcLiveData.value = result
+      }
+
+      RecipientEntity.Type.BCC -> {
+        recipientsBccLiveData.value = result
+      }
+    }
+  }
+
+  /**
+   * Get information about [RecipientWithPubKeys] from the remote server.
+   *
+   * @param email Used to generate a request to the server.
+   * @return [RecipientWithPubKeys]
+   * @throws IOException
+   */
+  private suspend fun getPgpInfoFromServer(
+    email: String? = null,
+    fingerprint: String? = null
+  ): List<PgpKeyDetails>? = withContext(Dispatchers.IO) {
```

Comment:
Here you should be able to remove the `fingerprint` argument. Make email not-nullable.

Also please rename the method to `getPublicKeysFromRemoteServers`

Reply:
Will do